src/main/java/com/google/sps/servlets/SearchServlet.java

```diff
@@ -40,91 +40,114 @@
+   */
+  private List<Recipe> recipesMatching(
+    HttpServletRequest request,
+    HttpServletResponse response,
+    String stringToMatch
+  )
+    throws ServletException, IOException {
     DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+    List<Recipe> recipesMatchingList = new ArrayList<>();
+    Index index = getIndex("recipes_index");
 
-    Query query = buildQuery(titleToMatch);
-    Results<ScoredDocument> resultsByTitle = searchIndex.search(query);
-    for (ScoredDocument entity : resultsByTitle) {
+    Query query = buildQuery(stringToMatch, 10);
+    Results<ScoredDocument> results = index.search(query);
+    for (ScoredDocument entity : results) {
       try {
         Entity recipeEntity = getRecipeEntityFromDocumentEntity(
           datastore,
           entity
         );
+
         Recipe recipe = buildRecipe(recipeEntity);
-        recipesList.add(recipe);
+        recipesMatchingList.add(recipe);
       } catch (EntityNotFoundException e) {
         response.setStatus(505);
         request.getRequestDispatcher("/search.jsp").forward(request, response);
-        return;
+        return new ArrayList<>();
       }
     }
-    request.setAttribute("recipesList", recipesList);
-    request.getRequestDispatcher("/search.jsp").forward(request, response);
+
+    return recipesMatchingList;
   }
 
   /**
    * Returns the index that stores the recipes documents
    */
-  private Index getIndex() {
-    IndexSpec indexSpec = IndexSpec
-      .newBuilder()
-      .setName("recipesIndex")
-      .build();
+  private Index getIndex(String indexName) {
+    IndexSpec indexSpec = IndexSpec.newBuilder().setName(indexName).build();
     Index index = SearchServiceFactory.getSearchService().getIndex(indexSpec);
     return index;
   }
 
-  private Query buildQuery(String titleToMatch) {
+  private Query buildQuery(String stringToMatch, int limit) {
     Cursor responseCursor = Cursor.newBuilder().build();
     QueryOptions options = QueryOptions
       .newBuilder()
-      .setLimit(10)
-      .setFieldsToSnippet("title")
-      .setCursor(responseCursor)
+      .setLimit(limit)
       .setSortOptions(
-        SortOptions
-          .newBuilder()
-          .addSortExpression(
-            SortExpression
-              .newBuilder()
-              .setExpression("title")
-              .setDirection(SortExpression.SortDirection.ASCENDING)
-              .setDefaultValue("")
-          )
+        SortOptions.newBuilder().setMatchScorer(MatchScorer.newBuilder())
       )
+      .setCursor(responseCursor)
       .build();
 
-    Query query = Query
-      .newBuilder()
-      .setOptions(options)
-      .build(titleToMatch.toLowerCase());
+    Query query = Query.newBuilder().setOptions(options).build(stringToMatch);
 
     return query;
   }
 
   /**
-   * Build a Recipe Object with the given Entity
+   * Build a Recipe Object with the given Entity.
    */
```

Comment:
It could be good to include the author, difficulty and cooking time  in the recipe object. So that you can display these info in the search result page under the recipe title and it gives the user a better overview of the recipe.